src/main/java/com/artozersky/HackerNewsAPI/service/impl/NewsPostServiceImpl.java

```diff
@@ -0,0 +1,262 @@
+        
+        responseDTO.setMessage("Post created successfully");
+
+        updateCacheWithTopPosts();
+        
+        return responseDTO;
+    }
+
+    @Override
+    public NewsPostsResponseDTOImpl updatePost(NewsPostsUpdateDTOImpl postUpdateDTO, Long postId) {
+        
+        NewsPostModelImpl post = postRepository.findById(postId)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + postId));
+        
+        boolean isRedundant = true;
+        
+        if (postUpdateDTO.getTitle() != null && !postUpdateDTO.getTitle().trim().isEmpty()) {
+            if (!postUpdateDTO.getTitle().equals(post.getTitle())) {
+                post.setTitle(postUpdateDTO.getTitle());
+                isRedundant = false;
+            }
+        }
+        
+        if (postUpdateDTO.getUrl() != null && !postUpdateDTO.getUrl().trim().isEmpty()) {
+            if (!postUpdateDTO.getUrl().equals(post.getUrl())) {
+                post.setUrl(postUpdateDTO.getUrl());
+                isRedundant = false;
+            }
+        }
+        
+        if(isRedundant) {
+            NewsPostsResponseDTOImpl responseDTO = modelMapper.map(post, NewsPostsResponseDTOImpl.class);
+            responseDTO.setMessage("No changes detected. Update skipped.");
+            return responseDTO;
+        }
+
+        post.onPostUpdate();
+        
+        NewsPostModelImpl updatedPost = postRepository.save(post);
+        
+        updateCacheWithTopPosts();
+
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(updatedPost, NewsPostsResponseDTOImpl.class);
+        responseDTO.setMessage("Post updated successfully");
+
+        
+        return responseDTO;
+    }
+    
+    @Override
+    public NewsPostsResponseDTOImpl upVote(Long id) {
+        NewsPostModelImpl post = postRepository.findById(id)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + id));
+        post.upVote();        
+        NewsPostModelImpl updatedPost = postRepository.save(post);
+
+        //if in the cache update the cache
+        if (cacheService.getPostFromCacheById(id) != null) {
+            cacheService.putPostInCache(updatedPost);
+        }
+        //updateCacheWithTopPosts();
+
+
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(updatedPost, NewsPostsResponseDTOImpl.class);
+
+        responseDTO.setMessage("Vote updated successfully");
+        
+        return responseDTO;
+    }
+    
+    @Override
+    public NewsPostsResponseDTOImpl downVote(Long id) {
+        NewsPostModelImpl post = postRepository.findById(id)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + id));
+        post.downVote();        
+        NewsPostModelImpl updatedPost = postRepository.save(post);
+
+        //downvote can make a post go down in the rank and out of sorted cache
+        updateCacheWithTopPosts();
```

Comment:
upvoted can't make posts change places and then need to resort the cache?

Reply:
sure